File: typescript/src/payment/application/policies/paymentRejection.ts
// Policy che traduce TransactionRejected in RejectTransaction.

import type { Policy } from '@/common/application/policy.js';
import { RejectTransaction, rejectTransaction } from '../../application/commands/rejectTransaction.js';
import { TransactionRejected } from '../../domain/events/transactionRejected.js';
import { requireArgument } from '@/common/utils/requireArgument.js';

export class PaymentRejection implements Policy<TransactionRejected, RejectTransaction> {
  evaluate(event: TransactionRejected): RejectTransaction {
    requireArgument(event, 'event');
    return rejectTransaction(event.aggregateId, event.transactionId, event.reason);
  }
}

File: java/src/main/java/it/giannibombelli/wsc2026/payment/application/policies/PaymentRejection.java
package it.giannibombelli.wsc2026.payment.application.policies;

import it.giannibombelli.wsc2026.common.utils.Require;

import it.giannibombelli.wsc2026.common.application.Policy;
import it.giannibombelli.wsc2026.payment.application.commands.RejectTransaction;
import it.giannibombelli.wsc2026.payment.domain.events.TransactionRejected;

public final class PaymentRejection implements Policy<TransactionRejected, RejectTransaction> {

    @Override
    public RejectTransaction evaluate(TransactionRejected event) {
        Require.requireArgument(event, "event");
        return new RejectTransaction(
            event.aggregateId(),
            event.transactionId(),
            event.reason()
        );
    }
}
